File: baekjoon/dp/9251/9251-lcs.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <algorithm>
#include <string>

using namespace std;
int d[1001][1001];


int main() {
	string a, b;
	cin >> a >> b;

	int m = a.size();
	int n = b.size();
	int ans = 0;

	for (int i = 1; i <= m; i++) {
		for (int j = 1; j <= n; j++) {
			if (a[i - 1] == b[j - 1]) {
				d[i][j] = d[i - 1][j - 1] + 1;
			}
			else {
				d[i][j] = max(d[i - 1][j], d[i][j - 1]);
			}
		}
	}
	cout << d[m][n] << endl;
	return 0;
}

File: baekjoon/dp/9251/9251-lcs.py
a = input()
b = input()

d = [[0] * (len(a) + 1) for _ in range(len(b) + 1)]
ans = 0

for i in range(1, len(b) + 1):
    for j in range(1, len(a) + 1):
        if b[i - 1] == a[j - 1]:
            d[i][j] = d[i - 1][j - 1] + 1
        else:
            d[i][j] = max(d[i][j - 1], d[i - 1][j])

        ans = max(ans, d[i][j])

for arr in d:
    print(arr)

print(ans)
